### PR TITLE
Add auto-discovery of relations for PostgresQL and SpatiaLite

### PR DESCRIPTION
**Pull request by pvalsecc**
Fixed the add relation functionnality: the table is sorted. When the code
was setting the sorted column, the row was sorted and the other columns it was
setting were set on the wrong row.

### CODE REVIEW

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 1]
```diff
@@ -3907,6 +3908,84 @@ QVariant QgsPostgresProvider::convertValue( QVariant::Type type, QVariant::Type
...
+    if ( pgProvider &&
+         pgProvider->mUri.connectionInfo( false ) == connectionInfo && pgProvider->mSchemaName == schema && pgProvider->mTableName == tableName )
+    {
+      result.append( layer );
+    }
+  }
+  return result;
+}
+
+QList<QgsRelation> QgsPostgresProvider::discoverRelations( const QgsVectorLayer* self, const QList<QgsVectorLayer*>& layers ) const
```
Comment (luipir) [inline src/providers/postgres/qgspostgresprovider.cpp]: as far as I can read, it could be applied also to other DB providers, isn't it?

File: src/core/qgsrelation.h [hunk 2]
```diff
@@ -272,6 +280,15 @@ class CORE_EXPORT QgsRelation
...
     bool isValid() const;
 
+    /**
+     * Compares the two QgsRelation, ignoring the name and the ID.
+     *
+     * @param other The other relation
+     * @return true if they are similar
+     * @note added in QGIS 3.0
+     */
+    bool isSame( const QgsRelation& other ) const;
```
Comment (3nids) [inline src/core/qgsrelation.h]: `isEquivalent` sounds better to my [french] ears... maybe a native speaker has a better phrasing?

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 1 above)
Comment (pvalsecc) [inline src/providers/postgres/qgspostgresprovider.cpp]: What do you mean? Discovering relations or this function?
For discovering relations, I plan to make another PR for sqlite.
For the searchLayers method, it could be make a template and kind of work for other providers. But I'd rather refactor it when I do the sqlite implementation.

File: src/core/qgsrelation.h (hunk 2 above)
Comment (pvalsecc) [inline src/core/qgsrelation.h]: What's the problem with `isSame`? I've seen in other APIs use it and I don't see why the method should have a name that works for both english and french...

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 1 above)
Comment (luipir) [inline src/providers/postgres/qgspostgresprovider.cpp]: exactly, I was talking about discovery relation... as I can read, the only db dependent part of all the complete PR, is this query => can be moved directly to sqlite. As you said will be part of another PR. thank you :)

File: src/core/qgsrelation.h (hunk 2 above)
Comment (nyalldawson) [inline src/core/qgsrelation.h]: It's really awkward naming. I'd also prefer isEquivalent or isEqual, or even better something which more clearly describes the function like hasEqualDefinition.

Comment (pvalsecc) [inline src/core/qgsrelation.h]: OK, renaming to `hasEqualDefinition`. And I'll rename the static function in `qgsrelationmanager.cpp` to `hasRelationWithEqualDefinition` while at it.

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 1 above)
Comment (pvalsecc) [inline src/providers/postgres/qgspostgresprovider.cpp]: I've added the SpatiaLite stuff to this PR since it's ready and the PR is still open.

**Comment by pvalsecc** [comment]
All comments are now addressed.

**Comment by pvalsecc** [comment]
@luipir, @3nids, do you have other concerns about his PR?

**Comment by luipir** [comment]
no concerns from me :)

**Comment by 3nids** [comment]
no more cosmetics on my side ;)
I have no expertise to provide better feedback, sorry.

**Comment by h-cas-valle** [comment]
I'm new to this, but as far as I understand, if it would be possible, this feature would be a life saver for teamworking with PosgreSQL databases
#27918
